src/orange/construct/DepthCalculator.hh

```diff
@@ -0,0 +1,54 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file orange/construct/DepthCalculator.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <unordered_map>
+#include <vector>
+
+#include "corecel/cont/VariantUtils.hh"
+
+#include "OrangeInput.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate the maximum number of levels deep in a geometry.
+ */
+class DepthCalculator
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using VecVarUniv = std::vector<VariantUniverseInput>;
+    //!@}
+
+  public:
+    // Construct with a reference to all universe inputs
+    explicit DepthCalculator(VecVarUniv const&);
+
+    //! Calculate the depth of the global unit
+    size_type operator()() { return this->visit_univ_(*this, 0); }
+
+    // Calculate the depth of a unit
+    size_type operator()(UnitInput const& u);
```

Comment:
The second two `operator()` functions can be made private

Reply:
Actually, it can't, because they get called by `std::visit`'s low-level implementaiton. It's similar to how you can't make a private constructor if you use `std::make_shared`.